src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.cs

```diff
@@ -863,13 +863,16 @@ private ImmutableHashSet<ISymbol> ComputeGeneratedCodeSymbolsInTree(SyntaxTree t
                     CompilationEvent e;
                     try
                     {
-                        if (!prePopulatedEventQueue)
+                        if (!CompilationEventQueue.TryDequeue(out e))
                         {
-                            e = await CompilationEventQueue.DequeueAsync(cancellationToken).ConfigureAwait(false);
-                        }
-                        else if (!CompilationEventQueue.TryDequeue(out e))
-                        {
-                            return completedEvent;
+                            if (!prePopulatedEventQueue)
+                            {
+                                e = await CompilationEventQueue.DequeueAsync(cancellationToken).ConfigureAwait(false);
```

Comment:
can we make AsyncQueue.DequeueAsync return a ValueTask instead? (and propagate that up here?)

Reply:
:arrow_right: I originally implemented this that way, but after review I felt it would unnecessarily delay the improvement.

Reply:
Can you link to that review?  I'm curious waht woudl be wrong with that approach.  Thanks!

Reply:
#23557

Reply:
Any reason we shouldn't pursue the ValueTask thing first? It's extra work but extra work that will pay dividends in the long run. I'd like to think that shipping a new façade that we can use to significantly improve perf would be an easy RPS exception to get.

Reply:
@jasonmalinowski now that we know `ValueTask<T>` is not just a type forwarding assembly, it may be harder to get it approved. I'm still pursuing that option but I don't think it should block this PR since it's an improvement either way.

Reply:
This is an internal API. Why not just copie the code for ValueTask and use it that way? No assembly load necessary.